specs/core/0_beacon-chain.md

```diff
@@ -1903,144 +1983,348 @@ def process_transfer(state: BeaconState, transfer: Transfer) -> None:
+        for crosslink_committee, shard in get_crosslink_committees_at_slot(state, slot):
+            winning_root, participants = get_winning_root_and_participants(state, shard)
+            participating_balance = get_total_balance(state, participants)
+            total_balance = get_total_balance(state, crosslink_committee)
+            if 3 * participating_balance >= 2 * total_balance:
+                state.latest_crosslinks[shard] = Crosslink(
+                    epoch=slot_to_epoch(slot),
+                    crosslink_data_root=winning_root,
+                )
+```
 
-#### Crosslinks
+#### Eth1 data
 
-For every `slot in range(get_epoch_start_slot(previous_epoch), get_epoch_start_slot(next_epoch))`, let `crosslink_committees_at_slot = get_crosslink_committees_at_slot(state, slot)`. For every `(crosslink_committee, shard)` in `crosslink_committees_at_slot`, compute:
+Run the following function:
 
-* Set `state.latest_crosslinks[shard] = Crosslink(epoch=slot_to_epoch(slot), crosslink_data_root=winning_root(crosslink_committee))` if `3 * total_attesting_balance(crosslink_committee) >= 2 * get_total_balance(crosslink_committee)`.
+```python
+def maybe_reset_eth1_period(state: BeaconState) -> None:
+    if (get_current_epoch(state) + 1) % EPOCHS_PER_ETH1_VOTING_PERIOD == 0:
+        for eth1_data_vote in state.eth1_data_votes:
+            # If a majority of all votes were for a particular eth1_data value,
+            # then set that as the new canonical value
+            if eth1_data_vote.vote_count * 2 > EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH:
+                state.latest_eth1_data = eth1_data_vote.eth1_data
+        state.eth1_data_votes = []
+```
 
 #### Rewards and penalties
 
 First, we define some additional helpers:
 
-* Let `base_reward_quotient = integer_squareroot(previous_total_balance) // BASE_REWARD_QUOTIENT`.
-* Let `base_reward(state, index) = get_effective_balance(state, index) // base_reward_quotient // 5` for any validator with the given `index`.
-* Let `inactivity_penalty(state, index, epochs_since_finality) = base_reward(state, index) + get_effective_balance(state, index) * epochs_since_finality // INACTIVITY_PENALTY_QUOTIENT // 2` for any validator with the given `index`.
-
-Note: When applying penalties in the following balance recalculations implementers should make sure the `uint64` does not underflow.
-
-##### Justification and finalization
+```python
+def get_base_reward(state: BeaconState, index: ValidatorIndex) -> Gwei:
+    if get_previous_total_balance(state) == 0:
+        return 0
 
-* Let `previous_active_validator_indices = get_active_validator_indices(state.validator_registry, previous_epoch)`
-* Let `epochs_since_finality = next_epoch - state.finalized_epoch`.
+    adjusted_quotient = integer_squareroot(get_previous_total_balance(state)) // BASE_REWARD_QUOTIENT
+    return get_effective_balance(state, index) // adjusted_quotient // 5
```

Comment:
do we want to do some constant folding upon the final parameter tuning for these numbers? same below w/ `INACTIVITY_PENALTY_QUOTIENT // 2` in `get_inactivity_penalty`